src/gstring.rs

```diff
@@ -0,0 +1,440 @@
+        self.as_str()
+    }
+}
+
+impl Ord for GString {
+    fn cmp(&self, other: &GString) -> Ordering {
+        self.as_str().bytes().cmp(other.as_str().bytes())
+    }
+}
+
+impl PartialOrd for GString {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for GString {
+    fn eq(&self, other: &GString) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<GString> for String {
+    fn eq(&self, other: &GString) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<str> for GString {
+    fn eq(&self, other: &str) -> bool {
+        self.as_str() == other
+    }
+}
+
+impl<'a> PartialEq<&'a str> for GString {
+    fn eq(&self, other: &&'a str) -> bool {
+        self.as_str() == *other
+    }
+}
+
+impl PartialEq<String> for GString {
+    fn eq(&self, other: &String) -> bool {
+        self.as_str() == other.as_str()
+    }
+}
+
+impl PartialEq<GString> for str {
+    fn eq(&self, other: &GString) -> bool {
+        self == other.as_str()
+    }
+}
+
+impl PartialOrd<GString> for String {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(&String::from(other.as_str())))
+    }
+}
+
+impl PartialOrd<String> for GString {
+    fn partial_cmp(&self, other: &String) -> Option<Ordering> {
+        Some(self.as_str().cmp(other.as_str()))
+    }
+}
+
+impl PartialOrd<GString> for str {
+    fn partial_cmp(&self, other: &GString) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+impl PartialOrd<str> for GString {
+    fn partial_cmp(&self, other: &str) -> Option<Ordering> {
+        Some(self.as_str().cmp(other))
+    }
+}
+
+impl Eq for GString {}
+
+impl AsRef<str> for GString {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl AsRef<OsStr> for GString {
+    fn as_ref(&self) -> &OsStr {
+        OsStr::new(self.as_str())
+    }
+}
+
+impl Deref for GString {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl From<GString> for String {
+    #[inline]
+    fn from(s: GString) -> Self {
+        std::mem::forget(s.ptr);
+        String::from(s.as_str())
+    }
+}
+
+impl From<GString> for Box<str> {
+    #[inline]
+    fn from(s: GString) -> Self {
+        std::mem::forget(s.ptr);
+        Box::from(s.as_str())
+    }
+}
+
+impl From<String> for GString {
+    #[inline]
+    fn from(s: String) -> Self {
+        GString::from(s.as_bytes().to_vec())
+    }
+}
+
+impl From<Box<str>> for GString {
+    #[inline]
+    fn from(s: Box<str>) -> Self {
+        s.as_bytes().to_vec().into()
+    }
+}
+
+impl From<&'static str> for GString {
+    #[inline]
+    fn from(s: &'static str) -> Self {
+        s.as_bytes().to_vec().into()
+    }
+}
+
+impl From<Vec<u8>> for GString {
+    #[inline]
+    fn from(s: Vec<u8>) -> Self {
+        let cstr = unsafe { CString::from_vec_unchecked(s) };
```

Comment:
This seems unsafe. You could create strings with this that are not NUL-terminated